### PR TITLE
Render Details contact page via BlocksBuilder

### PR DESCRIPTION
**Pull request by zpawn**
Story https://app.shortcut.com/deskpro/story/165035/hubspot-contact-field-mapping

### CODE REVIEW

File: src/components/Admin/ContactMapping/ContactMapping.tsx [hunk 1]
```diff
@@ -1,16 +1,48 @@
...
+                onClickTab={onChangeTab}
+                containerStyles={{
+                    padding: "0 12px",
+                    justifyContent: "flex-start",
+                    margin: "16px 0 8px",
+                    gap: "10px",
+
+                }}
+            />
+            {(activeTab === 0) && (
```
Comment (HappyPaul55) [inline src/components/Admin/ContactMapping/ContactMapping.tsx]: Don't use magic numbers, instead use consts, e.g. `activeTab === TAB_HOME`

File: src/utils/getScreenStructure.ts [hunk 2]
```diff
@@ -11,7 +11,11 @@ const getScreenStructure = (
     return match([object, screen])
         .with(["contact", "home"], () => settings?.mapping_contact
             ? JSON.parse(settings.mapping_contact)
-            : STRUCTURE.CONTACT
+            : STRUCTURE.CONTACT.home
+        )
+        .with(["contact", "view"], () => settings?.mapping_contact
```
Comment (HappyPaul55) [inline src/utils/getScreenStructure.ts]: Not sure how this `with` works. If it is `contact`, does it run all code blocks or just the first. Would an `if/else` be easier to follow with fewer dependancies?

**Review by HappyPaul55 — CHANGES_REQUESTED** [review]
Looks good overall, just a few comments.

File: src/components/Admin/ContactMapping/ContactMapping.tsx (hunk 1 above)
Comment (zpawn) [reply src/components/Admin/ContactMapping/ContactMapping.tsx]: fixed with constant: 
```js
const TABS = {
    HOME: 0,
    VIEW: 1,
};
```

File: src/utils/getScreenStructure.ts (hunk 2 above)
Comment (zpawn) [reply src/utils/getScreenStructure.ts]: We use `ts-pattern` in almost all projects. `match` works as follows:

In `match`, we pass the value that we want to check. The value can be of any type (primitive, array, or object). In `.with`, we check for a match and return a value.

It's more like a `switch-case` but more powerful and flexible.

Comment (HappyPaul55) [reply src/utils/getScreenStructure.ts]: It seems overkill and that vanilla typescript can be handled nicely with the return early pattern.

```typescript
    if (object === 'contact' && screen === 'home') {
        return settings?.mapping_contact
            ? JSON.parse(settings.mapping_contact)
            : STRUCTURE.CONTACT.home
    }
    if (object === 'contact' && screen === 'view') {
        return settings?.mapping_contact
            ? JSON.parse(settings.mapping_contact)
            : STRUCTURE.CONTACT.view
    }
    return [];
```

Or maybe even just
```typescript
    if (object === 'contact') {
        return settings?.mapping_contact
            ? JSON.parse(settings.mapping_contact)
            : (STRUCTURE.CONTACT[screen] ?? [])
    }
    return [];
```

Comment (zpawn) [reply src/utils/getScreenStructure.ts]: rewrote with first example

File: src/utils/getScreenStructure.ts [hunk 3]
```diff
@@ -8,12 +7,17 @@ const getScreenStructure = (
...
             ? JSON.parse(settings.mapping_contact)
-            : STRUCTURE.CONTACT
-        )
-        .otherwise(() => []);
+            : STRUCTURE.CONTACT.home;
+    } else if (object === "contact" && screen === "view") {
+        return settings?.mapping_contact
+            ? JSON.parse(settings.mapping_contact)
+            : STRUCTURE.CONTACT.view
+    } else {
```
Comment (HappyPaul55) [inline src/utils/getScreenStructure.ts:18]: No need for the `else`, just a `return` a the bottom will be fine.

File: src/utils/getScreenStructure.ts [hunk 4]
```diff
@@ -8,12 +7,17 @@ const getScreenStructure = (
...
-    return match([object, screen])
-        .with(["contact", "home"], () => settings?.mapping_contact
+    if (object === "contact" && screen === "home") {
+        return settings?.mapping_contact
             ? JSON.parse(settings.mapping_contact)
-            : STRUCTURE.CONTACT
-        )
-        .otherwise(() => []);
+            : STRUCTURE.CONTACT.home;
+    } else if (object === "contact" && screen === "view") {
```
Comment (HappyPaul55) [inline src/utils/getScreenStructure.ts:14]: No need for `else`. Just an if on the next line.

**Review by HappyPaul55 — APPROVED** [review]

**Pull request merged**